pkg/codegen/schema/schema.go

```diff
@@ -0,0 +1,898 @@
+type Type interface {
+	String() string
+
+	isType()
+}
+
+type primitiveType int
+
+const (
+	boolType    primitiveType = 1
+	intType     primitiveType = 2
+	numberType  primitiveType = 3
+	stringType  primitiveType = 4
+	archiveType primitiveType = 5
+	assetType   primitiveType = 6
+	anyType     primitiveType = 7
+)
+
+func (t primitiveType) String() string {
+	switch t {
+	case boolType:
+		return "boolean"
+	case intType:
+		return "integer"
+	case numberType:
+		return "number"
+	case stringType:
+		return "string"
+	case archiveType:
+		return "pulumi:pulumi:Archive"
+	case assetType:
+		return "pulumi:pulumi:Asset"
+	case anyType:
+		return "pulumi:pulumi:Any"
+	default:
+		panic("unknown primitive type")
+	}
+}
+
+func (primitiveType) isType() {}
+
+// IsPrimitiveType returns true if the given Type is a primitive type. The primitive types are bool, int, number,
+// string, archive, asset, and any.
+func IsPrimitiveType(t Type) bool {
+	_, ok := t.(primitiveType)
+	return ok
+}
+
+var (
+	// BoolType represents the set of boolean values.
+	BoolType Type = primitiveType(boolType)
+	// IntType represents the set of 32-bit integer values.
+	IntType Type = primitiveType(intType)
+	// NumberType represents the set of IEEE754 double-precision values.
+	NumberType Type = primitiveType(numberType)
+	// StringType represents the set of UTF-8 string values.
+	StringType Type = primitiveType(stringType)
+	// ArchiveType represents the set of Pulumi Archive values.
+	ArchiveType Type = primitiveType(archiveType)
+	// AssetType represents the set of Pulumi Asset values.
+	AssetType Type = primitiveType(assetType)
+	// AnyType represents the complete set of values.
+	AnyType Type = primitiveType(anyType)
+)
+
+// MapType represents maps from strings to particular element types.
+type MapType struct {
+	// ElementType is the element type of the map.
+	ElementType Type
+}
+
+func (t *MapType) String() string {
+	return fmt.Sprintf("Map<%v>", t.ElementType)
+}
+
+func (*MapType) isType() {}
+
+// ArrayType represents arrays of particular element types.
+type ArrayType struct {
+	// ElementType is the element type of the array.
+	ElementType Type
+}
+
+func (t *ArrayType) String() string {
+	return fmt.Sprintf("Array<%v>", t.ElementType)
+}
+
+func (*ArrayType) isType() {}
+
+// UnionType represents values that may be any one of a specified set of types.
+type UnionType struct {
+	// ElementTypes are the allowable types for the union type.
+	ElementTypes []Type
+}
+
+func (t *UnionType) String() string {
+	elements := make([]string, len(t.ElementTypes))
+	for i, e := range t.ElementTypes {
+		elements[i] = e.String()
+	}
+	return fmt.Sprintf("Union<%v>", strings.Join(elements, ", "))
+}
+
+func (*UnionType) isType() {}
+
+// ObjectType represents schematized maps from strings to particular types.
+type ObjectType struct {
+	// Token is the type's Pulumi type token.
+	Token string
+	// Comment is the description of the type, if any.
+	Comment string
```

Comment:
i thought i saw before some stuff about deprecation.  is that not needed?

Reply:
We don't deprecate nested types at the moment. We could add support for a deprecation message here if/when we need it..